include/rmm/mr/device/aligned_resource_adaptor.hpp

```diff
@@ -0,0 +1,194 @@
+    RMM_EXPECTS(allocation_alignment % 256 == 0, "Allocation alignment is not a multiple of 256.");
+    RMM_EXPECTS(alignment_threshold % 256 == 0, "Alignment threshold is not a multiple of 256.");
+  }
+
+  aligned_resource_adaptor()                                = delete;
+  ~aligned_resource_adaptor() override                      = default;
+  aligned_resource_adaptor(aligned_resource_adaptor const&) = delete;
+  aligned_resource_adaptor(aligned_resource_adaptor&&)      = delete;
+  aligned_resource_adaptor& operator=(aligned_resource_adaptor const&) = delete;
+  aligned_resource_adaptor& operator=(aligned_resource_adaptor&&) = delete;
+
+  /**
+   * @brief Get the upstream memory resource.
+   *
+   * @return Upstream* pointer to a memory resource object.
+   */
+  Upstream* get_upstream() const noexcept { return upstream_; }
+
+  /**
+   * @copydoc rmm::mr::device_memory_resource::supports_streams()
+   */
+  [[nodiscard]] bool supports_streams() const noexcept override
+  {
+    return upstream_->supports_streams();
+  }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return bool true if the upstream resource supports get_mem_info, false otherwise.
+   */
+  [[nodiscard]] bool supports_get_mem_info() const noexcept override
+  {
+    return upstream_->supports_get_mem_info();
+  }
+
+ private:
+  static constexpr std::size_t default_allocation_alignment = 256;
+  static constexpr std::size_t default_alignment_threshold  = 0;
+
+  /**
+   * @brief Allocates memory of size at least `bytes` using the upstream resource with the specified
+   * alignment.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   * by the upstream resource.
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param stream Stream on which to perform the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cuda_stream_view stream) override
+  {
+    if (allocation_alignment_ == default_allocation_alignment || bytes < alignment_threshold_) {
+      return upstream_->allocate(bytes, stream);
+    } else {
+      auto const aligned_size = rmm::detail::align_up(bytes, allocation_alignment_);
+      auto const size         = aligned_size + allocation_alignment_ - default_allocation_alignment;
+      auto const address      = reinterpret_cast<std::size_t>(upstream_->allocate(size, stream));
+      auto const aligned_address = rmm::detail::align_up(address, allocation_alignment_);
+      auto const head_size       = aligned_address - address;
+      auto const tail_size       = size - head_size - aligned_size;
+      if (head_size != 0) { upstream_->deallocate(reinterpret_cast<void*>(address), head_size); }
+      if (tail_size != 0) {
+        upstream_->deallocate(reinterpret_cast<void*>(aligned_address + aligned_size), tail_size);
+      }
```

Comment:
This is a violation of the `device_memory_resource` interface as it is calling `deallocate` on a pointer `p` with a different size than the `allocate` call that returned it. Furthermore, it is deallocating a pointer that was never returned from `allocate` in the first place.

Reply:
We used to track allocations in the pool and arena resources, but this has been turned off by default (#702, #732), so at least pool/arena can handle these types of deallocations. If we don't want to relax these constraints, we'd have to implement custom alignment in pool/arena themselves.

Reply:
We cannot relax these constraints. They are fundamental to the interface. 

Alignment can be done as an adaptor by using the `map<void*,void*>` approach I described above.

Reply:
I was hoping to keep this adapter stateless and not waste memory, but it turns out to be not too bad. PTAL